File: Aoc.Csharp/Solutions/Day09.cs
global using static Aoc.Runner.LogHelpers;
namespace Aoc.Solutions.Day09;

public class Map
{
    public Map(List<List<int>> heights) => Heights = heights;

    public List<List<int>> Heights { get; }
    public int Rows => Heights.Count;
    public int Cols => Heights[0].Count;

    public IEnumerable<(int r, int c)> SurroundingPoints(int r, int c)
    {
        foreach (var ri in Range(r - 1, 3))
            foreach (var ci in Range(c - 1, 3))
                if (ri >= 0 && ri < Rows && ci >= 0 && ci < Cols && (ci == c || ri == r) && !(ri == r && ci == c))
                    yield return (ri, ci);
    }

    public IEnumerable<int> Surrounding(int r, int c) => SurroundingPoints(r, c).Select(p => Heights[p.r][p.c]);

    public bool IsLow(int r, int c) =>
        Surrounding(r, c).All(h => h > Heights[r][c]);

    public int Risk(int r, int c) =>
        IsLow(r, c) ? Heights[r][c] + 1 : 0;

    public IEnumerable<(int r, int c)> Points()
    {
        foreach (var ri in Range(0, Rows))
            foreach (var ci in Range(0, Cols))
                yield return (ri, ci);
    }

    // get basin from starting low point
    public List<(int r, int c)> GetBasin(int r, int c)
    {
        Log($"\nLow point: {(r, c)}");
        var basin = new HashSet<(int r, int c)>() { (r, c) };
        var inBasin = (int ri, int ci) =>
            SurroundingPoints(ri, ci).Any(p => basin.Contains((p.r, p.c)));

        var ring = 1;
        var newMatches = true;
        while (newMatches)
        {
            Log($"Ring = {ring} - lastMatch = {newMatches}");
            newMatches = false;
            foreach (var ri in Range(r - ring, ring * 2 + 1))
            {
                foreach (var ci in Range(c - ring, ring * 2 + 1))
                {
                    if (ri >= 0 && ri < Rows && ci >= 0 && ci < Cols && Heights[ri][ci] != 9)
                    {
                        var anySurrounding = inBasin(ri, ci);
                        Log($"Checking {(ri, ci) } = {Heights[ri][ci]} - {anySurrounding }");
                        if (anySurrounding && !basin.Contains((ri, ci)))
                        {
                            Log($"\tAdding {(ri, ci)}");
                            basin.Add((ri, ci));
                            newMatches = true;
                        }
                    }
                }
            }
            ring += 1;
        }
        Log($"Basin from {(r, c)} - {basin.Select(b => b.ToString()).Join(",")}");
        return basin.ToList();
    }
}

public class Day09 : Day
{
    public static Map Parse(string inputString) =>
        inputString
            .Split('\n')
            .Select(line => line.Trim().Select(c => int.Parse(new string(new char[] { c }))).ToList())
            .ToList()
            .Then(lines => new Map(lines));

    public override string SolveA(string input)
    {
        var map = Parse(input);
        return map.Points().Select(point => map.Risk(point.r, point.c)).Sum().ToString();
    }

    public override string SolveB(string input)
    {
        var map = Parse(input);
        var lowPoints = map.Points().Where(p => map.IsLow(p.r, p.c)).ToList();
        var basins =
            lowPoints
            .Select(p => map.GetBasin(p.r, p.c))
            .ToList();
        return
            basins
            .Select(b => b.Count)
            .OrderByDescending(c => c)
            .Take(3)
            .Aggregate(1, (i, state) => i * state)
            .ToString();
    }

    public Day09()
    {
        Tests = new()
        {
            new("A", @"2199943210
3987894921
9856789892
8767896789
9899965678", "15", SolveA),
            new("B", @"2199943210
3987894921
9856789892
8767896789
9899965678", "1134", SolveB)
        };
    }
}

File: Aoc.Fsharp/Solutions/Day09.fs
module Aoc.Solutions.Day09

open Aoc.Runner
open System.Collections.Generic

let surrounding (r, c) (heights: int[][]) =
    let rows = Seq.length heights
    let cols = Seq.length heights[0]
    seq {
        for ri in r - 1 .. r + 1 do
            for ci in c - 1 .. c + 1 ->
                (ri, ci)
    }
    |> Seq.filter (fun (ri, ci) ->
                ri >= 0 && ri < rows && ci >= 0 && ci < cols && (ci = c || ri = r) && not(ri = r && ci = c))

let surroundingValue point heights =
    surrounding point heights
    |> Seq.map (fun (r, c) -> heights[r][c])

let isLow (r, c) heights =
    surrounding (r, c) heights
    |> Seq.forall (fun (ri, ci) ->
        heights[r][c] < heights[ri][ci])

let risk (r, c) heights =
    if isLow (r, c) heights then
        heights[r][c] + 1
    else
        0

let points (heights: int[][]) =
    seq {
        for r in 0 .. Seq.length heights - 1 do
            for c in 0 .. Seq.length heights[0] - 1 ->
                (r, c)
    }

let basin point (heights: int[][]) =
    let basin = new HashSet<int * int>()
    let rows = Seq.length heights
    let cols = Seq.length heights[0]
    let r, c = point
    basin.Add(point) |> ignore

    let inBasin somePoint =
        surrounding somePoint heights
        |> Seq.exists (fun another -> basin.Contains(another))

    let mutable ring = 1
    let mutable matches = true

    while matches = true do
        matches <- false
        for ri in r - ring .. r + ring + 1 do
            for ci in c - ring .. c + ring + 1 do
                if ci >= 0 && ci < cols && ri >= 0 && ri < rows && heights[ri][ci] <> 9 then
                    let anySurrounding = inBasin (ri, ci)
                    if anySurrounding && not(basin.Contains((ri, ci))) then
                        basin.Add((ri, ci)) |> ignore
                        matches <- true
        ring <- ring + 1

    basin

let inline charToInt c = int c - int '0'
let parse (input: string) =
    input.Split('\n')
    |> Array.map (fun line -> line.Trim() |> Seq.toArray |> Array.map charToInt)

type Day09() =
    inherit Day()

    override _.SolveA input =
        let heights = parse input
        heights
        |> points
        |> Seq.map (fun point -> risk point heights)
        |> Seq.sum
        |> string

    override _.SolveB input =
        let heights = parse input
        let lows = heights |> points |> Seq.filter (fun p -> isLow p heights)
        lows
        |> Seq.map (fun p -> basin p heights)
        |> Seq.map Seq.length
        |> Seq.sortDescending
        |> Seq.take 3
        |> Seq.fold (fun xs x -> x * xs) 1
        |> string